org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/CodeActionHandler.java

```diff
@@ -283,16 +285,30 @@ private void populateDataFields(List<Either<Command, CodeAction>> codeActions) {
 	}
 
 	public static IProblemLocationCore[] getProblemLocationCores(ICompilationUnit unit, List<Diagnostic> diagnostics) {
-		IProblemLocationCore[] locations = new IProblemLocationCore[diagnostics.size()];
+		List<IProblemLocationCore> locations = new ArrayList<>();
 		for (int i = 0; i < diagnostics.size(); i++) {
 			Diagnostic diagnostic = diagnostics.get(i);
-			int problemId = getProblemId(diagnostic);
-			int start = DiagnosticsHelper.getStartOffset(unit, diagnostic.getRange());
-			int end = DiagnosticsHelper.getEndOffset(unit, diagnostic.getRange());
-			boolean isError = diagnostic.getSeverity() == DiagnosticSeverity.Error;
-			locations[i] = new ProblemLocationCore(start, end - start, problemId, new String[0], isError, IJavaModelMarker.JAVA_MODEL_PROBLEM_MARKER);
+			if (!handledProblems(diagnostic, diagnostics)) {
```

Comment:
Is it possible to have this logic be done in https://github.com/eclipse/eclipse.jdt.ls/blob/master/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/corrections/QuickFixProcessor.java#L73 ? That's what I meant from my first comment.

You would need to pass the diagnostic data (the problem argument into the IProblemLocationCore argument data) but it would have the deduplication logic all in one spot.

Also, I would just add some Java comment above `if(!handledProbelms((..))` to explain we need to de-duplicate different problems with the same resolution, and reference the issue.

Reply:
Fixed.